src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -4,44 +4,257 @@
+            /// </summary>
+            private const int MaxLog2PrimitiveFieldSize = 3;
+
+            /// <summary>
+            /// <a href="https://github.com/dotnet/coreclr/blob/master/src/vm/class.h#L66">MAX_PRIMITIVE_FIELD_SIZE</a>
+            /// </summary>
+            private const int MaxPrimitiveFieldSize = 1 << MaxLog2PrimitiveFieldSize;
+
+            /// <summary>
+            /// CoreCLR DomainLocalModule::OffsetOfDataBlob() / sizeof(void *)
+            /// </summary>
+            private const int DomainLocalModuleDataBlobOffsetAsIntPtrCount = 6;
+
+            /// <summary>
+            /// CoreCLR ThreadLocalModule::OffsetOfDataBlob() / sizeof(void *)
+            /// </summary>
+            private const int ThreadLocalModuleDataBlobOffsetAsIntPtrCount = 3;
+
+            /// <summary>
+            /// TargetDetails is used to query pointer size for the target architecture.
+            /// </summary>
+            private TargetDetails _target;
+
+            public ModuleFieldLayoutMap(TargetDetails target)
+            {
+                _target = target;
+            }
+
+            protected override bool CompareKeyToValue(EcmaModule key, ModuleFieldLayout value)
+            {
+                return key == value.Module;
+            }
+
+            protected override bool CompareValueToValue(ModuleFieldLayout value1, ModuleFieldLayout value2)
+            {
+                return value1.Module == value2.Module;
+            }
+
+            protected override ModuleFieldLayout CreateValueFromKey(EcmaModule module)
+            {
+                int typeCountInModule = module.MetadataReader.GetTableRowCount(TableIndex.TypeDef);
+
+                // 0 corresponds to "normal" statics, 1 to thread-local statics
+                LayoutInt[] gcStatics = new LayoutInt[2]
+                {
+                    LayoutInt.Zero,
+                    LayoutInt.Zero
+                };
+                LayoutInt[] nonGcStatics = new LayoutInt[2]
+                {
+                    new LayoutInt(DomainLocalModuleDataBlobOffsetAsIntPtrCount * _target.PointerSize + typeCountInModule),
+                    new LayoutInt(ThreadLocalModuleDataBlobOffsetAsIntPtrCount * _target.PointerSize + typeCountInModule),
+                };
+                Dictionary<DefType, FieldAndOffset[]> typeToFieldMap = new Dictionary<DefType, FieldAndOffset[]>();
+
+                for (int typeIndex = 1; typeIndex <= typeCountInModule; typeIndex++)
+                {
+                    List<FieldAndOffset> fieldsForType = null;
+                    DefType defType = (DefType)module.GetObject(MetadataTokens.TypeDefinitionHandle(typeIndex));
+                    if (defType.HasInstantiation)
+                    {
+                        // Generic types are exempt from the static field layout algorithm, see
+                        // <a href="https://github.com/dotnet/coreclr/blob/master/src/vm/ceeload.cpp#L2049">this check</a>.
```

Comment:
These links into the CoreCLR codebase are going to be invalidated as soon as someone edits the file.

You can get a canonical link using the GitHub UI by hitting the `y` key on your keyboard. That one will point to a particular commit and won't get outdated.

Reply:
I have modified the two links in the file as you suggest.